src/platform/backends/shared_linux/process.cpp

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (C) 2019 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "process.h"
+#include "process_spec.h"
+
+#include <multipass/logging/log.h>
+
+namespace mp = multipass;
+namespace mpl = multipass::logging;
+
+mp::Process::Process(std::unique_ptr<mp::ProcessSpec>&& spec) : process_spec{std::move(spec)}
+{
+    setProgram(process_spec->program());
+    setArguments(process_spec->arguments());
+    setProcessEnvironment(process_spec->environment());
+    if (!process_spec->working_directory().isNull())
+        setWorkingDirectory(process_spec->working_directory());
+
+    setProcessChannelMode(QProcess::ForwardedErrorChannel); // default to forwarding child stderr to console
+}
+
+void mp::Process::start(const QStringList& extra_arguments)
+{
+    QProcess::start(program(), arguments() << extra_arguments);
+}
+
+bool mp::Process::run_and_return_status(const QStringList& extra_arguments, const int timeout)
+{
+    start(extra_arguments);
+    if (!waitForFinished(timeout))
```

Comment:
`waitForFinished` returns `false` if the process already finished, should we not check `exitStatus` here?

Reply:
Likely yes. I didn't try to improve the logic for this iteration, just move it around. But I'll improve it now, it's not a bad opportunity to do so.

Reply:
This confused me a little, so I looked into it. The docs for QProcess::waitForFinished do indeed say

> Returns true if the process finished; otherwise returns false (if the operation timed out, if an error occurred, or if this QProcess is already finished).

but the last clause is subtle.

QProcess lives on the thread of its instantiator, and the code it has internally for updating its known state of the child process, also lives on this thread. QProcess doesn't create a thread to monitor the child process it spawns, instead it creates [QSocketNotifiers](https://doc.qt.io/qt-5/qsocketnotifier.html) to monitors file descriptors of the process (stdin/out/err and an [FD for the process itself](https://www.macieira.org/blog/2012/07/forkfd-part-4-proposed-solutions/)) and emit signals on changes.

QSocketNotifiers docs state that: `File descriptor action is detected in Qt's main event loop`, i.e. a signal is emitted from this thread.

Here we are running
```
start();
waitForFinished();
```
one command after another. Say if the child process did run and quit, before waitForFinished was called. In that case, a signal is emitted on the Qt main loop (_q_processDied) to the QProcess, which will be processed on a following event loop tick.

Since we start and wait in the same function, there is no opportunity in between those two function calls for the event loop to tick, so the QProcess won't change its state from starting to finished just yet.

When we call waitForFinished(), QProcess will then actively poll the FDs until the process goes away (or finds it already has died), and the FD closes, and then update its state to finished.

So I think the way we use it is correct here.
- also, if the binary is missing, waitForFinished also returns false

If start() and waitForFinished() were called in different event loop ticks, it would be possible for the internal state to be finished before waitForFinished() is called (and so would return false). We should avoid using QProcess like that, and instead rely on the finished() signal.


But you are right that we should check exitStatus() to ensure the process didn't crash, as a crashing process will still have waitForFinished return true.